### PR TITLE
Prepare parts of the balanced kmeans for ivf-pq

### PR DESCRIPTION
**Pull request by achirkin**
Backport a piece of implementation from cuann/ivf-pq. This include:
  1. Move part of the internal kmeans code to GPU, thus reducing the number of necessary managed memory allocations;
  2. Split some of the functions into smaller reusable pieces;
  3. Move dataset-trainset sampling to the caller side.

### CODE REVIEW

**Comment by achirkin** [comment]
rerun tests

**Comment by achirkin** [comment]
rerun tests

**Review by tfeher — APPROVED** [review]
Thanks Artem for the PR, it is nice to get this piece on the GPU! I have two small comments, otherwise it looks good to me. Pre-approving.

**Comment by achirkin** [comment]
Tamas, thanks for the review! I'm re-requesting it, because I need to add more changes to reduce duplication and facilitate re-use in both ivf-pq and ivf-flat.

**Comment by achirkin** [comment]
rerun tests

File: cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh [hunk 1]
```diff
@@ -759,30 +808,30 @@ void build_optimized_kmeans(const handle_t& handle,
...
+  // Fine-tuning kmeans for all clusters
+  //
+  // (*) Since the likely cluster centroids have been calculated
+  // hierarchically already, the number of iteration for fine-tuning
+  // kmeans for whole clusters should be reduced. However, there
+  // is a possibility that the clusters could be unbalanced here,
+  // in which case the actual number of iterations would be increased.
+  //
+  balancing_em_iters(handle,
+                     std::max<uint32_t>(n_iters / 10, 2),
```
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh:822]: Is `n_iters / 10` the value that was used in IVF-PQ?

File: cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh [hunk 2]
```diff
@@ -421,32 +422,52 @@ auto adjust_centers(float* centers,
...
+ *
+ * @param handle
+ * @param n_iters the requested number of iteration
+ * @param dim the dimensionality of the dataset
+ * @param[in] dataset a pointer to a managed row-major array [n_rows, dim]
+ * @param n_rows the number of rows in the dataset
+ * @param n_cluster the requested number of clusters
+ * @param[inout] cluster_centers a pointer to a managed row-major array [n_clusters, dim]
+ * @param[out] cluster_labels a pointer to a managed row-major array [n_rows]
+ * @param[out] cluster_labels a pointer to a managed row-major array [n_clusters]
```
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh:442]:
```suggestion
 * @param[out] cluster_sizes a pointer to a managed row-major array [n_clusters]
```

File: cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh [hunk 3]
```diff
@@ -421,32 +422,52 @@ auto adjust_centers(float* centers,
...
+ * @param metric the distance type (there is a tweak in place for the similarity-based metrics)
+ * @param balancing_pullback
+ *   if the cluster centers are rebalanced on this number of iterations,
+ *   one extra iteration is performed (this could happen several times) (default should be `2`).
```
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh:446]: Could you reword it to make it clearer? This is what I understand:
- without any rebalancing (adjustment), we make n_iters iteration,
- each adjustment will add extra iterations,
- The number of the additional iterations is `ballancing_pullback-1`

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Artem for the update! Looks good, apart from a few minor clarification.

File: cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh (hunk 1 above)
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh:822]: Yes, and I left the comment from cuann repo as-is. I figure, this fine-tuning step isn't really specific to the ivf-pq use-case, so I've backported the changes here.

File: cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh (hunk 3 above)
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ann_kmeans_balanced.cuh:446]: Not exactly. The first and then every `ballancing_pullback`  rebalancing operations add one more iteration to the main cycle. I'll try to make it more understandable!

**Review by tfeher — APPROVED** [review]
Thanks @achirkin for the clarification. LGTM.

**Comment by cjnolet** [comment]
@gpucibot merge
